### PR TITLE
[RISCV] Allow crypto features to imply dependents

### PR DESCRIPTION
**Pull request by workingjubilee**
This relationship is a logical dependency.

Note Zvbc and Zvknhb. They are explicitly called out in the spec as requiring 64 bits:
- https://github.com/riscv/riscv-crypto/blob/56ed7952d13eb5bdff92e2b522404668952f416d/doc/vector/riscv-crypto-spec-vector.adoc

### CODE REVIEW

**Comment by workingjubilee** [comment]
This was brought to attention by https://github.com/rust-lang/rust/issues/131800

One could make some argument this is the job of the frontend, I think, and instead the frontend should _satisfy_ this dependency! ...but it would still be preferable to encode this relationship in the `.td`, so the dependency check can be generated.

**Comment by topperc** [comment]
> instead the frontend should _satisfy_ this dependency!

The .td file also generates code included by RISCVISAInfo.cpp used by clang to make -march imply the dependencies. This is needed to set preprocessor defines correctly.

File: llvm/lib/Target/RISCV/RISCVFeatures.td [hunk 1]
```diff
@@ -750,7 +751,8 @@ def HasStdExtZvbb : Predicate<"Subtarget->hasStdExtZvbb()">,
 
 def FeatureStdExtZvbc
     : RISCVExtension<"zvbc", 1, 0,
-                     "'Zvbc' (Vector Carryless Multiplication)">,
+                     "'Zvbc' (Vector Carryless Multiplication)",
+                     [FeatureStdExtZve64x]>,
       RISCVExtensionBitmask<0, 49>;
 def HasStdExtZvbc : Predicate<"Subtarget->hasStdExtZvbc()">,
                     AssemblerPredicate<(all_of FeatureStdExtZvbc),
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVFeatures.td:758]: FeatureStdExtZvbc32e below should imply Zve32x.

Comment (workingjubilee) [reply llvm/lib/Target/RISCV/RISCVFeatures.td:758]: Thank you!

**Comment by wangpc-pp** [comment]
Do we need to change `RISCVISAInfo::checkDependency()`?

**Review by lukel97 — COMMENTED** [review]
Does zvkt depend on zve32x?

**Comment by topperc** [comment]
> Does zvkt depend on zve32x?

It doesn't in gcc.

**Comment by workingjubilee** [comment]
According to the ISA spec so far, Zvkt doesn't imply *any* instructions. It thus doesn't imply that any register state must be addressable, either. It just specifies that other instructions should be modified in behavior if they are executed. Thus, perversely, Zvkt could be implemented on a processor that implements zero of the instructions that Zvkt modifies.

**Review by topperc — APPROVED** [review]
LGTM

**Comment by workingjubilee** [comment]
Not sure why it didn't whine when I ran it locally...?

**Comment by workingjubilee** [comment]
> Do we need to change `RISCVISAInfo::checkDependency()`?

...Probably? But because LLVM believes "negative features" are a sensible notion, I know there are some absolutely _wacky_ edge-cases in feature resolution which prevent me from feeling very confident in removing any code from that. The fact that in _normal_ usage the errors are simply naturally silenced by correctly resolving the features makes the 5 LOC that would be clawed back seem basically inconsequential.

Believe me, the thought of deleting more code greatly tempts me.

**Comment by topperc** [comment]
> > Do we need to change `RISCVISAInfo::checkDependency()`?
> 
> ...Probably? But because LLVM believes "negative features" are a sensible notion, I know there are some absolutely _wacky_ edge-cases in feature resolution which prevent me from feeling very confident in removing any code from that. The fact that in _normal_ usage the errors are simply naturally silenced by correctly resolving the features makes the 5 LOC that would be clawed back seem basically inconsequential.
> 
> Believe me, the thought of deleting more code greatly tempts me.

Let's go ahead and land this. I'll consider whether we can simplify `RISCVISAInfo::checkDependency()` separately.

**Pull request merged**